### PR TITLE
RavenDB-21180: Honor the request of Limit=0 & Offset=0 as a .count() func

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21180

### Additional description
Count operations should not consume more than work than just matching the documents and return the numbers found.

### Type of change
- Bug fix

### How risky is the change?
- Low

### CODE REVIEW

**Comment by arekpalinski** [comment]
Please check failing test: https://jenkins.hibernatingrhinos.com/job/PR_Tests/8078/

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 1]
```diff
@@ -646,6 +646,10 @@ private static bool WillAlwaysIncludeInResults(IndexType indexType, FieldsToFetc
                     long i = identityTracker.RegisterDuplicates(ref hasProjections, totalResults.Value, ids.AsSpan(0, read), token);
                     totalResults.Value += read; // important that this is *after* RegisterDuplicates
 
+                    // If we are going to just return count() then we don't care about anything else.
+                    if (query.IsCountQuery)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: That isn't sufficient for queries that want to get the total results as well, though
We have `break` below that stops us

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: We want to get total results for _any_ query, don't we?. This is about filling `QueryResult.TotalResults` with correct value. We always return it so they can use it e.g. for paging.

Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: This is a key, if we are going to always return a valid `QueryResult.TotalResults` then we need to handle this differently. We need to decide what do we expect from Corax.

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: We have to return a valid value there IMO. 

CC @ppekrol

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: @redknightlois For now we need to calculate it for each query and always return valid `QueryResult.TotalResults`.

We'll do the following detection on the client side:
https://issues.hibernatingrhinos.com/issue/RavenDB-21264/Detect-that-a-query-doesnt-require-to-return-statistics-QueryResult.TotalResults-in-particular
so in the future we'll be able to skip it.

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs [hunk 2]
```diff
@@ -122,11 +122,18 @@ protected static int CoraxBufferSize(global::Corax.IndexSearcher.IndexSearcher s
...
+            // just request the max buffer size instead.
+            if (numberOfEntries > DefaultBufferSizeForCorax)
                 return MaxBufferSizeForCorax;
             
             return DefaultBufferSizeForCorax;
         }
+
+        // Since we will need to evaluate the query completely anyways, it is preferable to err on the
+        // side of using extra memory to avoid repeated work. 
+        if (query.SkipStatistics || query.IsCountQuery)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs]: Should be:

```suggestion
        if (query.SkipStatistics == false || query.IsCountQuery)
```

**Comment by arekpalinski** [comment]
> Please check failing test: https://jenkins.hibernatingrhinos.com/job/PR_Tests/8078/

They are still failing - https://jenkins.hibernatingrhinos.com/job/PR_Tests/8151/

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 3]
```diff
@@ -726,11 +718,21 @@ private static bool WillAlwaysIncludeInResults(IndexType indexType, FieldsToFetc
...
                     if (docsToLoad <= 0)
                         break;
                 }
 
+
+                // If we are going to just return count() then we don't care about anything else than memoize the results.
+                if (query.IsCountQuery || query.SkipStatistics == false)
+                {
+                    var results = new MemoizationMatchProvider<IQueryMatch>(IndexSearcher, queryMatch);
+                    totalResults.Value += results.FillAndRetrieve().Length;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Can we dispose this immediately?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 4]
```diff
@@ -635,15 +626,16 @@ private static bool WillAlwaysIncludeInResults(IndexType indexType, FieldsToFetc
                             break;
                     }
                 }
-                
-                while (true)
+
+                // We don't need to do any processing for the query beyond counting if we are getting a count.
+                while (query.IsCountQuery == false)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: why `IsCountQuery==false`, is that changed in the loop?
I assume you want to say "if it's count only, don't do any work", but should move that to a variable to make it clearer

Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: The idea is to never enter into the loop. The cost of checking that is negligible, that's why I put the comment in there.

**Comment by arekpalinski** [comment]
Please check test failures: https://jenkins.hibernatingrhinos.com/job/PR_Tests/8193/

**Review by ayende — APPROVED** [review]

**Pull request merged**